pkg/cache/inmemory.go

```diff
@@ -0,0 +1,302 @@
+		return float64(c.maxSizeBytes)
+	})
+	_ = promauto.With(reg).NewGaugeFunc(prometheus.GaugeOpts{
+		Name: "thanos_cache_inmemory_max_item_size_bytes",
+		Help: "Maximum number of bytes for single entry to be held in the inmemory cache.",
+	}, func() float64 {
+		return float64(c.maxItemSizeBytes)
+	})
+
+	// Initialize LRU cache with a high size limit since we will manage evictions ourselves
+	// based on stored size using `RemoveOldest` method.
+	l, err := lru.NewLRU(maxInt, c.onEvict)
+	if err != nil {
+		return nil, err
+	}
+	c.lru = l
+
+	level.Info(logger).Log(
+		"msg", "created in-memory inmemory cache",
+		"maxItemSizeBytes", c.maxItemSizeBytes,
+		"maxSizeBytes", c.maxSizeBytes,
+		"maxItems", "maxInt",
+	)
+	return c, nil
+}
+
+func (c *InMemoryCache) onEvict(key, val interface{}) {
+	keySize := uint64(len(key.(string)))
+	entrySize := uint64(len(val.(cacheDataWithTTLWrapper).data))
+
+	c.evicted.Inc()
+	c.current.Dec()
+	c.currentSize.Sub(float64(entrySize))
+	c.totalCurrentSize.Sub(float64(keySize + entrySize))
+
+	c.curSize -= entrySize
+}
+
+func (c *InMemoryCache) get(key string) ([]byte, bool) {
+	c.requests.Inc()
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	v, ok := c.lru.Get(key)
+	if !ok {
+		return nil, false
+	}
+	// If the present time is greater than the TTL for the object from cache, the object will be
+	// removed from the cache and a nil will be returned
+	if time.Now().After(v.(cacheDataWithTTLWrapper).expiryTime) {
+		c.hitsExpired.Inc()
+		c.lru.Remove(key)
+		return nil, false
+	}
+	c.hits.Inc()
+	return v.(cacheDataWithTTLWrapper).data, true
+}
+
+func (c *InMemoryCache) set(key string, val []byte, ttl time.Duration) {
+	var size = uint64(len(val))
+	keySize := uint64(len(key))
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	if _, ok := c.lru.Get(key); ok {
+		return
+	}
+
+	if !c.ensureFits(size) {
+		c.overflow.Inc()
+		return
+	}
+
+	// The caller may be passing in a sub-slice of a huge array. Copy the data
+	// to ensure we don't waste huge amounts of space for something small.
+	v := make([]byte, len(val))
+	copy(v, val)
+	c.lru.Add(key, cacheDataWithTTLWrapper{data: v, expiryTime: time.Now().Add(ttl)})
+
+	c.added.Inc()
+	c.currentSize.Add(float64(size))
+	c.totalCurrentSize.Add(float64(keySize + size))
+	c.current.Inc()
+	c.curSize += size
+}
+
+// ensureFits tries to make sure that the passed slice will fit into the LRU cache.
+// Returns true if it will fit.
+func (c *InMemoryCache) ensureFits(size uint64) bool {
+	if size > c.maxItemSizeBytes {
+		level.Debug(c.logger).Log(
+			"msg", "item bigger than maxItemSizeBytes. Ignoring..",
+			"maxItemSizeBytes", c.maxItemSizeBytes,
+			"maxSizeBytes", c.maxSizeBytes,
+			"curSize", c.curSize,
+			"itemSize", size,
+		)
+		return false
+	}
+
+	for c.curSize+size > c.maxSizeBytes {
+		if _, _, ok := c.lru.RemoveOldest(); !ok {
+			level.Error(c.logger).Log(
+				"msg", "LRU has nothing more to evict, but we still cannot allocate the item. Resetting cache.",
```

Comment:
If there is nothing more to evict, it means that cache is already empty. Why doing the extra reset in that case?

Once we get to this point, we will never break out of this for-loop.

Reply:
I disagree that there is a infinite for loop. After `evict`ing all items and `reset`ing the cache, the condition `c.curSize+size > c.maxSizeBytes` will be `false` right? Cause `c.curSize == 0` and `size` cannot be greater than `maxSizeBytes` (line 95, 252). So there will no be an infinite loop right? 

Although I do get your point about the cache `reset` being unnecessary here. I looked into https://github.com/thanos-io/thanos/blob/1fff9a7157d1f3b7d5e0ce36f999b10bb3ae7ec4/pkg/store/cache/inmemory.go#L237 to write this function. Are you suggesting that even the `cache/inmemory.go#L237` is a bit incorrect too?

Reply:
> I disagree that there is a infinite for loop. After `evict`ing all items and `reset`ing the cache, the condition `c.curSize+size > c.maxSizeBytes` will be `false` right? Cause `c.curSize == 0` and `size` cannot be greater than `maxSizeBytes` (line 95, 252). So there will no be an infinite loop right?

You are right. I missed the check on line 95, and thought it would loop if `size > c.maxSizeBytes`. (Btw, unit-test for this scenario would be worth adding).

Reply:
> Are you suggesting that even the `cache/inmemory.go#L237` is a bit incorrect too?

Yes, I think `reset` there is not necessary, but harmless. (As in this case)

Reply:
>  (Btw, unit-test for this scenario would be worth adding).

Got it. :+1: